src/Lecture3.hs

```diff
@@ -213,8 +233,22 @@ types that can have such an instance.
 -- instance Foldable Weekday where
 -- instance Foldable Gold where
 -- instance Foldable Reward where
--- instance Foldable List1 where
--- instance Foldable Treasure where
+instance Foldable Treasure where
+  foldr :: (a -> b -> b) -> b -> Treasure a -> b
+  foldr f zero (SomeTreasure x) = f x zero
+  foldr _ zero NoTreasure       = zero
+
+  foldMap :: (Monoid m) => (a -> m) -> Treasure a -> m
+  foldMap _ NoTreasure       = mempty
+  foldMap f (SomeTreasure x) = f x
+
+instance Foldable List1 where
+  foldr :: (a -> b -> b) -> b -> List1 a -> b
+  foldr f zero (List1 x [])     = f x zero
+  foldr f zero (List1 x (y:ys)) = f y (foldr f zero (List1 x ys)) 
```

Comment:
You don't need to allocate a new `List1` on every new recursive call and you can reuse the `Foldable` instance for `[]` (and the `foldr` implementation specifically):

```suggestion
  foldr f zero (List1 x xs) = f x (foldr f zero xs)
```